### PR TITLE
[CS2113T-T09-4] CLICK

### PR DESCRIPTION
**Pull request by nvbinh15**
CLICK is a desktop app for managing modules, lifestyle, schedule, and CAP planning, optimized for use via a Command Line Interface (CLI).

### CODE REVIEW

File: docs/DeveloperGuide.md [hunk 1]
```diff
@@ -1,38 +1,432 @@
...
+accessing the packages and filtering out the commands, the `ClassPackageReader` presents the name of the command
+and the syntax in a readable format to the user. Do note that the packages have to be manually input by the developers.
+However, the core functionality of Click is already partitioned nicely into the packages and hence we do not expect
+many updates over the lifeline of this project.
+
+#### 4.6.2 Logic of Help Command
+
+After describing the [architecture](#461-architecture--of-help-command) of the help command, this portion will then describe the sequence of activation by
+the user when parsing a `help` command. Take the following sequence diagram for reference.
+![](./images/help/HelpCommand_execute-Help_Commands.png)
```
Comment (powzx) [inline docs/DeveloperGuide.md]: image

Should this arrow for execute() start before the activation bar ends?

File: docs/DeveloperGuide.md [hunk 2]
```diff
@@ -1,38 +1,432 @@
...
+
+| \ | Alternative 1 (current choice): Reads the name and syntax from the Classes | Alternative 2 (previous choice): Prints all available commands from a String, hard-coding every syntax and printing |
+|---|---|---|
+| Pros | 1. Dynamic, works well and sorts the names by order as long as the constructor is included for a command<br>2. Very readable and testable due to sorted names<br>3. OOP implementation with overloaded methods and branching on inheritance<br>4. The user gets to easily view *ALL*  possible commands with a single word | 1. Easy to implement, just adding all available commands into a String and print it out<br>2. Relatively fewer lines of code (LoC)<br>3. User gets specific syntax with command entered |
+| Cons | 1. Possible depreciated methods (`Class.getMethod`, `Class.getDeclaredConstructor`) which may be outdated, however,<br>  are still functional<br>2. Many more lines of code  (LoC) for implementation <br>3. The user  is bombarded with *ALL* possible commands with a single word | 1. Hard-coding and sorting help commands manually is a chore<br>2. User still has to remember the command in order to access the syntax |
```
Comment (giterator) [inline docs/DeveloperGuide.md]: designConsiderationsHelpCommand

File: docs/DeveloperGuide.md (hunk 1 above)
Comment (giterator) [inline docs/DeveloperGuide.md]: helpCommandDescription

Comment (giterator) [inline docs/DeveloperGuide.md]: helpCommandSequence

File: docs/DeveloperGuide.md [hunk 3]
```diff
@@ -1,38 +1,432 @@
...
+##### Sequence diagram when food is parsed
+
+The following diagram displays the interactions between the classes when the user enters a command starting with
+"food". You should take note of the interactions between the constructed command classes, and the current list it's
+iterating over - `WhatIAteTodayList`, especially the updates shown after the  end of every  command. An update is 
+defined as a manual overwrite over the text file saved in the user's hard disk. By convention, this text file is saved
+in the directory `fooddata` , with the text file named aptly as `food.txt`. Current improvements in this functionality
+would be to integrate the dates in `Calendar` with  the `food.txt` file, enabling the user to search what they ate on a
+given day.
+![](./images/food/food_architecture.png)
```
Comment (powzx) [inline docs/DeveloperGuide.md]: image

Is "Food Record" a method call? Should that arrow be starting before the activation bar ends?

Comment (powzx) [inline docs/DeveloperGuide.md]: image

Should "Food Record" start a new activation bar in AddFoodCommand? Should the activation for the AddFoodCommand be separated into multiple activation bars?

File: docs/DeveloperGuide.md [hunk 4]
```diff
@@ -1,38 +1,432 @@
...
+
+      iv.The method `arrangeLectureList` is also called and the process is same as in the previous step, except with `storage.lectureList` and `calendarLecture` replacing the first two input parameters of `arrangeTaskList`.
+      v.Then, the method `displayCalendar` in `Schedule` class is called, and it takes in `inputYearMonth` (the YearMonth object created from the month and year parsed from the user input), the `calendarTasks` (that was filled with the tasks for each day in the Step (iii)) and `calendarLecture` (that was filled with the lectures for each day in the Step (iv)). The method `displayCalendar` performs the necessary logic to print out the calendar.
+  >  **NOTE:** Two tasks and two lectures are displayed for each day based on the order in which the user added them, and if there are more, they will show as and when the user deletes the tasks/lectures that are currently displayed.
+
+The below sequence diagram shows the execution process of the calendar display feature.
+![](./images/calendar/CalendarDisplaySequence.png)
```
Comment (giterator) [inline docs/DeveloperGuide.md]: calendarDisplay1

Comment (giterator) [inline docs/DeveloperGuide.md]: calendarDisplay2

File: docs/DeveloperGuide.md [hunk 5]
```diff
@@ -1,38 +1,432 @@
...
+This section is designed to demonstrate our software design description, and aims to provide you with an overall guidance to the architecture of Duke.
+The following sequence diagram illustrates a command call by the user to Click. 
+The steps are as follows:
+1. When the user runs Click, a greeting would be printed to the user
+1. User enters some input which may be a command
+1. The command entered by the user would be parsed
+    1. if command, go ahead and execute it respectively
+    1. if not throw exception and ask for input again
+1. Parser returns control to Click
+![](./images/ClickRun.png)
```
Comment (giterator) [inline docs/DeveloperGuide.md]: commandCall

File: docs/DeveloperGuide.md [hunk 6]
```diff
@@ -1,38 +1,432 @@
...
+   ii. `Parser` calls `parser.parseCommand(userInput)` to parse user's input into a `Command`.
+2. Creating `ListModuleCommand` object.
+3. Executing command.\
+   i. `ListModuleCommand` calls `storage.StorageModule.readDataFromFile()` to read Module-related data from the storage file.\
+   ii. `ListModuleCommand` check if there is any Modules in the list. If not, prints the message of having no Modules then return.\
+   iii. `ListModuleCommand` prompts the message to list the Modules to the user and prints out the Modules line by line.
+
+The sequence diagram below summarizes how listing modules work:
+
+![](images/module/ListModule.png)
```
Comment (powzx) [inline docs/DeveloperGuide.md]: image

Should there also be a condition specified for the second alt block, e.g. [else]?

File: docs/DeveloperGuide.md [hunk 7]
```diff
@@ -1,38 +1,432 @@
...
+
+##### Class Diagram of Food
+
+Food-related commands operate on a list of food records, and a food storage object. The following diagram  illustrates
+how the storage in a text file, and the current food list interact with each other.
+Food storage object contains static methods to read and write data from said list.
+The saving of data from the list to the storage file is elaborated further in the diagram below.
+Further discussion on the design considerations of writing an abstract class are discussed in 
+design considerations.
+![](./images/food/foodClassDiagram.png)
```
Comment (powzx) [inline docs/DeveloperGuide.md]: image

Should there be a colon between isSilent and boolean?

Comment (powzx) [inline docs/DeveloperGuide.md]: image

Is the "C" icon standard notation?

Comment (giterator) [inline docs/DeveloperGuide.md]: foodDiag

File: docs/DeveloperGuide.md [hunk 8]
```diff
@@ -1,38 +1,432 @@
...
+be considered a simple feature. An interesting point for you to note before exploring this portion 
+would be the runtime-analysis of classes  done by the compiler, which reads the classes from the
+source code and extracts the syntax. This is perhaps more functional than printing out a string
+concatenating all command syntax.
+
+#### 4.6.1 Architecture  of Help Command
+
+The following diagram illustrates the  interactions between the functional class `ClassPackageReader` and
+a particular command `ClickCommand`.
+![](./images/help/helpCommandsClassDiagram.png)
```
Comment (powzx) [inline docs/DeveloperGuide.md]: image

Similar issue as before, is the "C" icon standard notation?

File: docs/DeveloperGuide.md [hunk 9]
```diff
@@ -1,38 +1,432 @@
...
+The steps are as follows:
+1. When the user runs Click, a greeting would be printed to the user
+1. User enters some input which may be a command
+1. The command entered by the user would be parsed
+    1. if command, go ahead and execute it respectively
+    1. if not throw exception and ask for input again
+1. Parser returns control to Click
+![](./images/ClickRun.png)
+
+You should note that this is a general overview of the Click functionality, and the `:Command` entity simply represents a Command to be called by the Parser.
```
Comment (uosjapuelks) [inline docs/DeveloperGuide.md:106]: With reference to the screenshot posted by giterator, your Click calls the command instead of your Parser.

File: docs/DeveloperGuide.md [hunk 10]
```diff
@@ -1,38 +1,432 @@
...
+1. User executes `module add c/CS2113T n/Software Engineering e/A`\
+   i. `Click` receives user's input.\
+   ii. `Parser` calls `parser.parseCommand(userInput)` to parse user's input into a `Command`.
+2. Creating `AddModuleCommand` object.
+3. Executing command.\
+   i. `AddModuleCommand` find `indexOfCode`, `indexOfName`, and `indexOfExpectedGrade` in user's input.\
+   ii. `AddModuleCommand` calls `getModule()` to create a new `module` based on user's input.\
+   iii. `AddModuleCommand` calls `storage.StorageModule.readDataFromFile()` to read Module-related data `moduleList` from the storage file.\
+   iv. `AddModuleCommand` calls `moduleList.addModule(module)` to add a new `module` to the list.\
+   v. `AddModuleCommand` prompts the successful message to the user.
```
Comment (uosjapuelks) [inline docs/DeveloperGuide.md:222]: image

You forgot to add a slash at the end this time so there's no new line.

File: docs/DeveloperGuide.md [hunk 11]
```diff
@@ -1,38 +1,432 @@
...
+1. User executes `module delete 2`\
+   i. `Click` receives user's input.\
+   ii. `Parser` calls `parser.parseCommand(userInput)` to parse user's input into a `Command`.
+2. Creating `DeleteModuleCommand` object.
+3. Executing command.\
+   i. `DeleteModuleCommand` calls `storage.StorageModule.readDataFromFile()` to read Module-related data from the storage file.\
+   ii. `DeleteModuleCommand` finds the `moduleIndex` based on user's input.\
+   iii. `DeleteModuleCommand` checks if `moduleIndex` is valid or not. If not, throw an `IllegalModuleIndexException`.\
+   iv. `DeleteModuleCommand` calls `moduleList.addModule(module)` to add a new `module` to the list.\
+   v. `DeleteModuleCommand` prompts the successful message to the user.\
```
Comment (uosjapuelks) [inline docs/DeveloperGuide.md]: image
You may not have intended to print successful message after adding a new module in "Delete" command.

File: docs/DeveloperGuide.md [hunk 12]
```diff
@@ -1,38 +1,432 @@
...
+      >  **NOTE:** The input is validated first and if the input date given is invalid, i.e., the month not between 1-12, then the calendar for the current month is displayed.
+   
+      ii. The `inputYearMonth` is passed into `Ui` class method `printCalenderTitle` and this prints out the title of that month with the month name and the year. In this example, it will display as given in the figure below.
+     ![](./images/calendar/calendar_header.png)
+      iii. Then, the method `arrangeTaskList` in `Schedule` class is called, and it takes in `storage.tasksList` (the TaskList object with all the currently stored tasks drawn from storage), `calendarTasks` (an ArrayList<ArrayList<String>> object initialized with empty ArrayLists of type String), `month` (the month input by the user, which in this example is the integer `10`) and year `month` (the year input by the user, which in this example is the integer `2021`), and adds the tasks to the days in the empty String ArrayLists initialized before in `calendarTasks`.
```
Comment (uosjapuelks) [inline docs/DeveloperGuide.md:359]: image
Almost thought point `iii.` disappeared into thin air. Do take note it appeared behind your screenshot.

**Review by uosjapuelks — COMMENTED** [review]
minor errors.

File: docs/DeveloperGuide.md [hunk 13]
```diff
@@ -1,38 +1,453 @@
...
+The following sequence diagram illustrates a command call by the user to Click. 
+The steps are as follows:
+1. When the user runs Click, a greeting would be printed to the user
+1. User enters some input which may be a command
+1. The command entered by the user would be parsed
+    1. if command, go ahead and execute it respectively
+    1. if not throw exception and ask for input again
+1. Parser returns control to Click
+![](./images/ClickRun.png)
+
```
Comment (naijie2108) [inline docs/DeveloperGuide.md:105]: image

Shouldn't methods include () behind the method?

File: docs/DeveloperGuide.md [hunk 14]
```diff
@@ -1,38 +1,453 @@
...
+Syntax: `zoom list`
+
+#### 4.2.2 Adding a new zoom link
+
+This command is implemented by the `AddZoomCommand` class. The basic functionality of this command is to write a new zoom link to a local storage file and associate it to the relevant module code. The `AddZoomCommand`
+class extends `Command`.
+
+Syntax: `zoom add LINK MODULE_CODE`
+Example: `zoom add nus.sg/testlink ABC101`
+
```
Comment (naijie2108) [inline docs/DeveloperGuide.md:338]: You can consider adding sequence diagrams, or more detailed explanation on how the zoom links are stored.

File: docs/DeveloperGuide.md (hunk 12 above)
Comment (naijie2108) [reply docs/DeveloperGuide.md:359]: Adding on to this, you can consider using markdown numbered lists.

File: docs/DeveloperGuide.md [hunk 15]
```diff
@@ -1,38 +1,453 @@
...
+and the syntax in a readable format to the user. Do note that the packages have to be manually input by the developers.
+However, the core functionality of Click is already partitioned nicely into the packages and hence we do not expect
+many updates over the lifeline of this project.
+
+#### 4.6.2 Logic of Help Command
+
+After describing the [architecture](#461-architecture--of-help-command) of the help command, this portion will then describe the sequence of activation by
+the user when parsing a `help` command. Take the following sequence diagram for reference.
+![](./images/help/HelpCommand_execute-Help_Commands.png)
+The sequence diagram provides a high-level view on how the entities interact. You should notice the interaction between
```
Comment (naijie2108) [inline docs/DeveloperGuide.md:871]: image
Is help a method parameter? If so, you should include the method as well

File: docs/DeveloperGuide.md [hunk 16]
```diff
@@ -1,38 +1,453 @@
...
+accessing the packages and filtering out the commands, the `ClassPackageReader` presents the name of the command
+and the syntax in a readable format to the user. Do note that the packages have to be manually input by the developers.
+However, the core functionality of Click is already partitioned nicely into the packages and hence we do not expect
+many updates over the lifeline of this project.
+
+#### 4.6.2 Logic of Help Command
+
+After describing the [architecture](#461-architecture--of-help-command) of the help command, this portion will then describe the sequence of activation by
+the user when parsing a `help` command. Take the following sequence diagram for reference.
+![](./images/help/HelpCommand_execute-Help_Commands.png)
```
Comment (naijie2108) [inline docs/DeveloperGuide.md]: image
Is "static" correct convention? I believe you should use "class" when calling static methods.

File: docs/DeveloperGuide.md [hunk 17]
```diff
@@ -1,38 +1,453 @@
...
+4. Click `Import Project`.
+5. Locate the `buid.gradle` file in the repository you have cloned and select it. Click `OK`.
+6. Click `OK` to accept the default setting.
+7. Verify the setup:\
+i. Go to `src/main/java/seedu.duke.Click`, run `Click.main()` and trying a few commands.\
+ii. Go to `src/test/java/seedu.duke` and run `Tests in seedu.duke` to ensure they all pass.
+
+## 3. Design
+
+This section is designed to demonstrate our software design description, and aims to provide you with an overall guidance to the architecture of Duke.
```
Comment (JMattChiam) [inline docs/DeveloperGuide.md]: Should it be "architecture of Click" instead?

File: docs/DeveloperGuide.md [hunk 18]
```diff
@@ -1,38 +1,453 @@
...
+The following sequence diagram illustrates a command call by the user to Click. 
+The steps are as follows:
+1. When the user runs Click, a greeting would be printed to the user
+1. User enters some input which may be a command
+1. The command entered by the user would be parsed
+    1. if command, go ahead and execute it respectively
+    1. if not throw exception and ask for input again
+1. Parser returns control to Click
+![](./images/ClickRun.png)
```
Comment (JMattChiam) [inline docs/DeveloperGuide.md]: splitCommandAndArgs method call should have an activation bar

Comment (JMattChiam) [inline docs/DeveloperGuide.md]: image

Should the execute() call directly into the activation bar of the Command constructor? shouldn't the constructor return the Object first, then execute called with its own activation bar?

File: docs/DeveloperGuide.md [hunk 19]
```diff
@@ -1,38 +1,453 @@
...
 
-## Design & implementation
+## 1. Introduction
 
-{Describe the design and implementation of the product. Use UML diagrams and short code snippets where applicable.}
+### 1.1 Background
 
+**_Welcome to Click_!** It's a Java-based  Command-Line Interface for Cramming and Knowledge (CLICK), providing a one-stop access point
+for managing various parts of your Computing student life here at NUS. We aim to provide a simple interface 
+that quantifies how you use your time and if you're taking care of your overall well-being.
```
Comment (JMattChiam) [inline docs/DeveloperGuide.md]: I think a brief overview of the application's capabilities and functionalities would be beneficial to understanding the rest of the DG easier.

File: docs/DeveloperGuide.md [hunk 20]
```diff
@@ -1,38 +1,453 @@
...
+
+This feature allows user to add a new Module.
+
+If the creation is successful, a confirmation message on the newly created Module will be displayed to the user.
+
+**Implementation**
+
+The command for listing all modules is implemented by the `AddModuleCommand` class that extends `Command`.
+
+When the user types `module add c/CS2113T n/Software Engineering e/A`, the following sequence of steps will then occur:
```
Comment (JMattChiam) [inline docs/DeveloperGuide.md]: For this and the other implementation sections, it would be good to explain what the different flags represent. Otherwise it's a bit hard to follow

File: docs/team/DG_binh.md [hunk 21]
```diff
@@ -0,0 +1,73 @@
+# Module-related Features
+
+## Adding a Module
+
+This feature allows user to add a new Module. 
+
+If the creation is successful, a confirmation message on the newly created Module will be displayed to the user.
+
+**Implementation**
```
Comment (leyondlee) [inline docs/team/DG_binh.md]: Can consider using UML diagram to illustrate this implementation

File: docs/team/DG_binh.md [hunk 22]
```diff
@@ -0,0 +1,73 @@
...
+ii. `Parser` calls `parser.parseCommand(userInput)` to parse user's input into a `Command`.
+2. Creating `ListModuleCommand` object.
+3. Executing command.\
+i. `ListModuleCommand` calls `storage.StorageModule.readDataFromFile()` to read Module-related data from the storage file.\
+ii. `ListModuleCommand` check if there is any Modules in the list. If not, prints the message of having no Modules then return.\
+iii. `ListModuleCommand` prompts the message to list the Modules to the user and prints out the Modules line by line.
+
+The sequence diagram below summarizes how listing modules work:
+
+![](../images/module/ListModule.png)
```
Comment (leyondlee) [inline docs/team/DG_binh.md]: image

print(...) self invocation seems to be incorrect.
Refer to: https://nus-cs2113-ay2122s1.github.io/website/se-book-adapted/chapters/uml.html#self-invocation

File: docs/DeveloperGuide.md [hunk 23]
```diff
@@ -1,38 +1,453 @@
...
+
+### 4.3 Calendar-related Features
+
+#### 4.3.1 Displaying the calendar
+
+**Implementation**
+
+The command for displaying the calendar for a specific month is implemented by the `DisplayCommand` class that extends `Command`.
+
+Given below is an example usage scenario and how the display calendar mechanism behaves at each step.
```
Comment (JMattChiam) [inline docs/DeveloperGuide.md]: This whole section is quite wordy with long paragraphs. It's quite hard to read and follow. 

Maybe putting the Sequence diagram at the front, then explaining the diagram might be clearer

File: docs/team/DG_swati.md [hunk 24]
```diff
@@ -0,0 +1,53 @@
...
+The method `arrangeLectureList` is also called and the process is same, except with `storage.lectureList` and `calendarLecture` replacing the first two input parameters of `arrangeTaskList`.
+- Then, the method `displayCalendar` in `Schedule` class is called, and it takes in `inputYearMonth` (the YearMonth object created from the month and year parsed from the user input), and `calendarTasks` (that was filled with the tasks for each day in the previous step). The method `displayCalendar` performs the necessary logic to print out a calendar with tasks as below.
+    >  **NOTE:** Two tasks and two lectures are displayed for each day based on the order in which the user added them, and if there are more, they will show as and when the user deletes the tasks/lectures that are currently displayed.
+
+The below sequence diagram shows the execution process of the calendar display feature.
+![](../images/calendar/CalendarDisplaySequence.png)
```
Comment (leyondlee) [inline docs/team/DG_swati.md]: image

1. Diagram is difficult to read due to scaling.
2. Return control arrows should be dotted. (Refer to: https://nus-cs2113-ay2122s1.github.io/website/se-book-adapted/chapters/uml.html#basic-2)

File: docs/DeveloperGuide.md (hunk 16 above)
Comment (JMattChiam) [inline docs/DeveloperGuide.md]: image

Shouldn't the activation bar for the parser call extend all the way until the help message is printed? 

Activation bars for Help command should be split up by the method calls. As it is now, it seems to imply that all the method calls occur in the constructor.

File: docs/DeveloperGuide.md (hunk 6 above)
Comment (leyondlee) [reply docs/DeveloperGuide.md]: image

Can indicate the return type for clarity instead of just `"moduleList"`.
